File: src/interview/designPattern/single/Singleton.java
package interview.designPattern.single;

public class Singleton {
    private Singleton() {
    }

    private static volatile Singleton instance;

    public static Singleton getInstance() {
        if (null == instance) {
            synchronized (Singleton.class) {
                if (null == instance) {
                    instance = new Singleton();
                }
            }
        }


        return instance;
    }



}

File: src/interview/designPattern/single/Singleton.kt
package interview.designPattern.single

class SingletonK private constructor() {
//    public var value: SingletonK? = null
    private object mHolder {
        val INSTANCE = SingletonK()
    }
    companion object Factory {
        fun getInstance(): SingletonK {
            return mHolder.INSTANCE
        }
    }
}